### PR TITLE
Add Synapse antispam module

### PR DESCRIPTION
**Pull request by turt2live**
This is looking for general code review rather than linting. Specifically, is this something that the backend team would be happy to run on matrix.org and is it valid enough Python?

**Requires https://github.com/matrix-org/synapse/pull/6259**

### CODE REVIEW

**Comment by erikjohnston** [comment]
Sorry, can you outline the functionality that this adds and use cases that this solves please?

**Comment by turt2live** [comment]
added to the README (hopefully).

File: synapse_antispam/mjolnir/ban_list.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+logger = logging.getLogger("synapse.contrib." + __name__)
+
+class BanList(object):
+    def __init__(self, api, room_id):
+        self.api = api
+        self.room_id = room_id
+        self.server_rules = []
+        self.user_rules = []
+        self.room_rules = []
+        self.build()
```
Comment (erikjohnston) [inline synapse_antispam/mjolnir/ban_list.py]: Note that this won't run synchronously, so the fields won't necessarily be up to date when things get called. I think we would probably rather want to have async getters here that fetched the data on demand (relying on caches to make it cheap), which has the added benefit of meaning we don't need to restart synapse to pick up a change in ban lists

Comment (turt2live) [reply synapse_antispam/mjolnir/ban_list.py]: we also rebuild lists over time as we see changes, so a restart shouldn't be required. This also isn't meant as a perfect solution, just something to catch 99.9% of the noise - a few seconds on startup waiting for a query isn't the end of the world imo.

We can make it more lazy, though it starts to affect message send times. The behaviour here is sacrificing a few events getting through in favour of being faster.

**Comment by turt2live** [comment]
I'm not sure it's worth documenting on the repo itself, but matrix.org's use case is to prevent undesirable servers from inviting our users to rooms. Of our two ban lists, we'd probably subscribe the homeserver to the terms of service ban list, therefore preventing the undesirable servers from performing invites. We do not want to subscribe matrix.org to the code of conduct ban list as it'll unnecessarily affect users who haven't violated the terms of service (and the code of conduct ban list is only meant to be applied in +matrix rooms, not the whole homeserver).

**Comment by erikjohnston** [comment]
This is probably fine for now. Once we get some more head room I'd like us to make it lazy just to avoid general confusion over whether a filter is working or a stuck cache or something. If we're spawning deferred without waiting on them then we need to wrap those places with the appropriate log context magic though

**Comment by turt2live** [comment]
> If we're spawning deferred without waiting on them then we need to wrap those places with the appropriate log context magic though

This part I'm unsure of how to do, but hoping to find it. 

---

Something else that occurred to me about lazily loading them: the API offered by Synapse doesn't use defer, so we'd be in a similar situation as to now.

File: synapse_antispam/mjolnir/ban_list.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+                logger.info("Adding rule %s/%s with action %s" % (event_type, entity, recommendation))
+                rule = ListRule(entity=entity, action=recommendation, reason=reason, kind=event_type)
+                if event_type in USER_RULE_TYPES:
+                    self.user_rules.append(rule)
+                elif event_type in ROOM_RULE_TYPES:
+                    self.room_rules.append(rule)
+                elif event_type in SERVER_RULE_TYPES:
+                    self.server_rules.append(rule)
+
+        with LoggingContext("mjolnir_ban_list_build"):
```
Comment (richvdh) [inline synapse_antispam/mjolnir/ban_list.py]: this is going to leak logging contexts.

I think you want `run_as_background_process`

**Pull request merged**